src/reconcile/rec_write.c

```diff
@@ -2043,6 +2044,220 @@ __rec_compression_adjust(WT_SESSION_IMPL *session, uint32_t max, size_t compress
+                LF_SET(WT_DELTA_HAS_STOP_TS);
+                WT_ERR(__wt_vpack_uint(&p, 0, supd->onpage_tombstone->start_ts));
+            }
+
+            if (supd->onpage_tombstone->durable_ts != WT_TS_NONE) {
+                LF_SET(WT_DELTA_HAS_STOP_DURABLE_TS);
+                WT_ERR(__wt_vpack_uint(&p, 0, supd->onpage_tombstone->durable_ts));
+            }
+        }
+
+        WT_ERR(__wt_vpack_uint(&p, 0, key->size));
+        memcpy(p, key->data, key->size);
+        p += key->size;
+
+        /* TODO: resolve modifies. Currently it works only for full values. */
+        WT_ERR(__wt_vpack_uint(&p, 0, supd->onpage_upd->size));
+        memcpy(p, supd->onpage_upd->data, supd->onpage_upd->size);
+        p += supd->onpage_upd->size;
+    }
+
+    r->delta.size += WT_PTRDIFF(p, head);
+    *head = flags;
+
+    WT_ASSERT(session, p < head + max_packed_size);
+err:
+    __wt_scr_free(session, &key);
+    return (ret);
+}
+
+/*
+ * __rec_build_delta_leaf --
+ *     Build delta for leaf pages.
+ */
+static int
+__rec_build_delta_leaf(WT_SESSION_IMPL *session, WT_RECONCILE *r)
+{
+    WT_DELTA_HEADER *header;
+    WT_MULTI *multi;
+    WT_SAVE_UPD *supd;
+    uint64_t start, stop;
+    uint32_t i;
+
+    WT_ASSERT(session, r->multi_next == 1);
+
+    start = __wt_clock(session);
+
+    multi = &r->multi[0];
+
+    WT_RET(__rec_build_delta_init(session, r));
+
+    for (i = 0, supd = multi->supd; i < multi->supd_entries; ++i, ++supd) {
+        if (supd->onpage_upd == NULL)
+            continue;
+
+        if (supd->onpage_tombstone != NULL && F_ISSET(supd->onpage_tombstone, WT_UPDATE_DURABLE))
+            continue;
+
+        if (supd->onpage_tombstone == NULL && F_ISSET(supd->onpage_upd, WT_UPDATE_DURABLE))
+            continue;
+
+        WT_RET(__rec_pack_delta_leaf(session, r, supd));
+    }
+
+    header = (WT_DELTA_HEADER *)r->delta.data;
+    header->mem_size = (uint32_t)r->delta.size;
+    header->type = r->ref->page->type;
+
+    /* TODO: write the delta to cloud. */
+
+    /* We cannot fail from here. */
+    for (i = 0, supd = multi->supd; i < multi->supd_entries; ++i, ++supd) {
+        if (supd->onpage_upd == NULL)
+            continue;
+
+        if (supd->onpage_tombstone != NULL) {
+            if (F_ISSET(supd->onpage_tombstone, WT_UPDATE_DURABLE))
+                continue;
+
+            F_SET(supd->onpage_tombstone, WT_UPDATE_DURABLE);
+        }
+
+        if (!F_ISSET(supd->onpage_upd, WT_UPDATE_DURABLE))
+            F_SET(supd->onpage_upd, WT_UPDATE_DURABLE);
+    }
+
+    stop = __wt_clock(session);
+
+    __wt_verbose(session, WT_VERB_PAGE_DELTA,
+      "Generated leaf page delta, original page size %d, delta size %d, "
+      "total time %" PRIu64 "us",
+      (int)r->ref->page->dsk->mem_size, (int)r->delta.size, WT_CLOCKDIFF_US(stop, start));
+
+    return (0);
+}
+
+/*
+ * __rec_build_delta --
+ *     Build delta.
+ */
+static int
+__rec_build_delta(WT_SESSION_IMPL *session, WT_RECONCILE *r)
```

Comment:
What are the preconditions for this? Do we expect it to be called for all pages, or just for leaf pages? If it should only be called for some pages, I think it makes sense to have an assertion here.

Reply:
It will be called for all pages in then end. But now only leaf page is supported.